src/main/java/uk/ac/ebi/embl/gff3tools/validation/builtin/FeatureSpecificValidation.java

```diff
@@ -0,0 +1,154 @@
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package uk.ac.ebi.embl.gff3tools.validation.builtin;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import uk.ac.ebi.embl.gff3tools.exception.ValidationException;
+import uk.ac.ebi.embl.gff3tools.gff3.GFF3Annotation;
+import uk.ac.ebi.embl.gff3tools.gff3.GFF3Attributes;
+import uk.ac.ebi.embl.gff3tools.gff3.GFF3Feature;
+import uk.ac.ebi.embl.gff3tools.utils.ConversionUtils;
+import uk.ac.ebi.embl.gff3tools.utils.OntologyClient;
+import uk.ac.ebi.embl.gff3tools.utils.OntologyTerm;
+import uk.ac.ebi.embl.gff3tools.validation.Validation;
+import uk.ac.ebi.embl.gff3tools.validation.meta.Gff3Validation;
+import uk.ac.ebi.embl.gff3tools.validation.meta.ValidationMethod;
+import uk.ac.ebi.embl.gff3tools.validation.meta.ValidationType;
+
+@Gff3Validation
+public class FeatureSpecificValidation extends Validation {
+
+    private static final String SINGLE_OPERON_MESSAGE =
+            "Feature \"%s\" refers to operon \"%s\". Please provide an operon feature which spans the entire operon region. Refer to (http://www.ebi.ac.uk/ena/WebFeat/operon_s.html) for details";
+    private static final String MULTIPLE_OPERON_MESSAGE =
+            "\"%s\" number of features refer to operon \"%s\". Please provide an operon feature which spans the entire operon region. Refer to (http://www.ebi.ac.uk/ena/WebFeat/operon_s.html) for details";
+
+    private final OntologyClient ontologyClient = ConversionUtils.getOntologyClient();
+
+    @ValidationMethod(rule = "OPERON_FEATURE", type = ValidationType.ANNOTATION)
+    public void validateOperonFeatures(GFF3Annotation annotation, int line) throws ValidationException {
+
+        boolean hasOperonFeature = false;
+        List<GFF3Feature> featuresWithOperon = new ArrayList<>();
+
+        for (GFF3Feature feature : annotation.getFeatures()) {
+            Optional<String> soIdOpt = ontologyClient.findTermByNameOrSynonym(feature.getName());
+            if (soIdOpt.isPresent()
+                    && (OntologyTerm.OPERON.ID.equals(soIdOpt.get())
+                            || ontologyClient.isChildOf(soIdOpt.get(), OntologyTerm.OPERON.ID))) {
+                hasOperonFeature = true;
+                break;
+            }
+            if (feature.hasAttribute(GFF3Attributes.OPERON)) {
+                featuresWithOperon.add(feature);
+            }
+        }
+
+        if (hasOperonFeature || featuresWithOperon.isEmpty()) {
+            return;
+        }
+
+        Map<String, List<GFF3Feature>> groupedByOperon = featuresWithOperon.stream()
+                .filter(f -> f.getAttributeByName(GFF3Attributes.OPERON) != null)
+                .collect(Collectors.groupingBy(f -> f.getAttributeByName(GFF3Attributes.OPERON)));
+
+        for (Map.Entry<String, List<GFF3Feature>> entry : groupedByOperon.entrySet()) {
```

Comment:
why do the double loop? Just do the validation on line 52 instead of accumulating them and then reinterating.